napari/plugins/hook_specifications.py

```diff
@@ -317,3 +322,35 @@ def napari_write_vectors(path: str, data: Any, meta: dict) -> Optional[str]:
         If data is successfully written, return the ``path`` that was written.
         Otherwise, if nothing was done, return ``None``.
     """
+
+
+# -------------------------------------------------------------------------- #
+#                                 GUI Hooks                                  #
+# -------------------------------------------------------------------------- #
+
+
+@napari_hook_specification(historic=True)
+def napari_experimental_provide_function_widget() -> Union[
+    MagicFunctionArg, List[MagicFunctionArg]
+]:
```

Comment:
What I dislike about aliases is that I have to hunt around for the definition, so I liked Talley's proposal of putting it next to the hookspec. If it needs to be used in multiple places, two options are:
- redefinition, which I think should work?
- including it just above the function in a comment.

One particular concern is how this will render. For example, if you look at our current hookspec docs:

https://napari.org/docs/dev/plugins/hook_specifications.html

you'll see some complicated-to-parse signatures *but* everything about those signatures is standard and easy to work through:

![hookspec-type-annot](https://user-images.githubusercontent.com/492549/104110572-8c794400-529e-11eb-817c-fcb810a39e17.png)

Do we want the docs to have only `MagicFunctionArg` there? Is there a standard tool that will expand that type alias on hover, or link aliases to their definition?

Given all this, I think probably the best way to do this is copy the alias in the docstring. Thoughts?

Reply:
This one is tricky.

I guess my ideal would be that `MagicFunctionArg` would be clickable in the docs and so you'd be able to click on it and then see what it is, and we'd only really have to define it in one place, and you wouldn't have to hunt around.

Without additional description something like
```python
Optional[Callable[[str, List[Tuple[Any, Dict, str]]], List[str]]]
```
is basically meaningless anyway. I don't think anyone could look at that and know what they are supposed to return without reading a lot more. I also struggle to parse all the nested `]` and so when I look at just have a really hard time with it.

I find the following much easier to parse
```python
Optional[Callable[[str, List[LayerDataTuple]], List[str]]]
```
It also makes it clear that `Tuple[Any, Dict, str]` in both `napari_get_writer` and `napari_get_reader` are the same thing.

Given by preference for the type aliases I'd probably just add the definition again in a comment above the code. I don't really like redefinition as they could then get out of sync.

Reply:
> Given by preference for the type aliases I'd probably just add the definition again in a comment above the code. I don't really like redefinition as they could then get out of sync.

Well, the nice thing is, if they get out of sync programmatically, our CI can catch it. If they get out of sync in the comments, then that is much harder to catch.

Also, adding the definition above the code doesn't solve the docs issue...

> I find the following much easier to parse
>
> Optional[Callable[[str, List[LayerDataTuple]], List[str]]]

A very good point! I'm on board that hover (click is not quite as friendly imho) on data type definitions is the ideal case. That might require some sphinx changes, but actually I'm seeing that they are either there or coming!

https://github.com/sphinx-doc/sphinx/pull/8007

A [GitHub code search for 'autodoc_type_aliases'](https://github.com/search?q=autodoc_type_aliases&type=code) took me here, where they seem to have correctly worked out how to link type aliases to their definition

https://pycgdescent.readthedocs.io/en/latest/#pycgdescent.minimize

The cool thing is that the definition includes a comment about how the alias works! 🤯

We should also alias something like `Pathlike = Union[str, pathlib.Path]`, I think.

One more thing, about the name: MagicFunctionArg seems vague, whereas LayerDataTuple feels concrete. How about AnnotatedFunction, GuiAnnotatedFunction, FunctionPlusGui, ...? Any other thoughts on this? Since this *is* now user-facing, I'd like to spend just a little bit of time coming up with a good name for it.

Reply:
> One more thing, about the name: MagicFunctionArg seems vague, whereas LayerDataTuple feels concrete. How about AnnotatedFunction, GuiAnnotatedFunction, FunctionPlusGui, ...? Any other thoughts on this? Since this is now user-facing, I'd like to spend just a little bit of time coming up with a good name for it.

Those are all possibilities, `AnnotatedFunction` might be my favourite, but I don't have strong feelings. We should think about what we do for the widget too. Could do `DockFunction` and `DockWidget` or `DockFunctionWidget`?

Reply:
> One particular concern is how this will render. For example, if you look at our current hookspec docs you'll see some complicated-to-parse signatures but everything about those signatures is standard and easy to work through:

this remains a challenge with sphinx.  Type aliases (as opposed to `NewType`) really are currently just internal shorteners for the things they alias (though @jni ... https://github.com/sphinx-doc/sphinx/pull/8007 looks good!). I have used [https://github.com/agronholm/sphinx-autodoc-typehints](sphinx-autodoc-typehints) to at least make the signatures look better in docs (leaving the verbose explanation to the docstring).  Since our docs are still very much in flux, maybe it's not critical for this PR how it looks in the docs?

> Since this is now user-facing, I'd like to spend just a little bit of time coming up with a good name for it.

I agree, names are important.  worth getting right.  I think the function itself should just be `Callable`.  The `Union` between that callable and the `Tuple` with additional `dicts` is really what we need to decide.  I picked `...Arg` cause those are all valid "arguments" for our hook registration process.  I'm not sure anything with the word `Function` in it is appropriate for something where a `tuple` is also valid.

Reply:
> Since our docs are still very much in flux, maybe it's not critical for this PR how it looks in the docs?

Yup, agreed.

> I agree, names are important. worth getting right. 

Yup. Logic for `...Arg` was sound to me. How about `DockFunctionArgs` and `DockWidgetArgs` for our two things respectively (I might lean towards `Args` a little over `Arg`, but would be fine with `Arg` if people thought that was more accurate).

Reply:
Or actually, even better, let's just go with the works in the hook spec itself, so

```python
def napari_experimental_provide_function_widget() -> Union[
    FunctionWidgetArgs, List[FunctionWidgetArgs]
]:
```
and
```python
def napari_experimental_provide_dock_widget() -> Union[
    DockWidgetArgs, List[DockWidgetArgs]
]:
```

Reply:
singular might be better than plural too, I don't mind

Reply:
See here for discussion: https://napari.zulipchat.com/#narrow/stream/270578-reviews/topic/2080/near/222245561 (requires registration/login with github). Summary:

- choice is not that critical — turns out to effectively only be a documentation choice, which can be reversed later. As such, this choice should not hold up merging of this PR.
- plural Args is better than singular Arg, as it more closely reflects that there is some unpacking to be done (even though it is not standard unpacking)
- all three of us gave the ok to FunctionWidgetArgs/DockWidgetArgs, *or* AnnotatedFunction/AnnotatedWidget, and all three of us liked a new option, AugmentedFunction/AugmentedWidget, which avoids the connotation/ambiguity as to whether these are type annotations or something else (or both).

This can get merged with or without the name change to `Augmented...`

Reply:
Renamed